homeassistant/components/zha/climate.py

```diff
@@ -118,708 +42,124 @@ async def async_setup_entry(
-
-        if (running_state := self._thrm.running_state) is None:
+        if self.entity_data.entity.hvac_action is None:
             return None
-        if running_state & (
-            T.RunningState.Heat_State_On | T.RunningState.Heat_2nd_Stage_On
-        ):
-            return HVACAction.HEATING
-        if running_state & (
-            T.RunningState.Cool_State_On | T.RunningState.Cool_2nd_Stage_On
-        ):
-            return HVACAction.COOLING
-        if running_state & (
-            T.RunningState.Fan_State_On
-            | T.RunningState.Fan_2nd_Stage_On
-            | T.RunningState.Fan_3rd_Stage_On
-        ):
-            return HVACAction.FAN
-        if running_state & T.RunningState.Idle:
-            return HVACAction.IDLE
-        if self.hvac_mode != HVACMode.OFF:
-            return HVACAction.IDLE
-        return HVACAction.OFF
-
-    @property
-    def _pi_demand_action(self) -> HVACAction | None:
-        """Return the current HVAC action based on pi_demands."""
-
-        heating_demand = self._thrm.pi_heating_demand
-        if heating_demand is not None and heating_demand > 0:
-            return HVACAction.HEATING
-        cooling_demand = self._thrm.pi_cooling_demand
-        if cooling_demand is not None and cooling_demand > 0:
-            return HVACAction.COOLING
-
-        if self.hvac_mode != HVACMode.OFF:
-            return HVACAction.IDLE
-        return HVACAction.OFF
+        return HVACAction(self.entity_data.entity.hvac_action)
 
     @property
     def hvac_mode(self) -> HVACMode | None:
         """Return HVAC operation mode."""
-        return SYSTEM_MODE_2_HVAC.get(self._thrm.system_mode)
+        if self.entity_data.entity.hvac_mode is None:
+            return None
+        return HVACMode(self.entity_data.entity.hvac_mode)
 
     @property
     def hvac_modes(self) -> list[HVACMode]:
         """Return the list of available HVAC operation modes."""
-        return SEQ_OF_OPERATION.get(self._thrm.ctrl_sequence_of_oper, [HVACMode.OFF])
+        return [HVACMode(mode) for mode in self.entity_data.entity.hvac_modes]
 
     @property
     def preset_mode(self) -> str:
         """Return current preset mode."""
-        return self._preset
+        return self.entity_data.entity.preset_mode
 
     @property
     def preset_modes(self) -> list[str] | None:
         """Return supported preset modes."""
-        return self._presets
+        return self.entity_data.entity.preset_modes
 
     @property
     def supported_features(self) -> ClimateEntityFeature:
         """Return the list of supported features."""
-        features = self._supported_flags
-        if HVACMode.HEAT_COOL in self.hvac_modes:
-            features |= ClimateEntityFeature.TARGET_TEMPERATURE_RANGE
-        if self._fan is not None:
-            self._supported_flags |= ClimateEntityFeature.FAN_MODE
-        return features
+        return ClimateEntityFeature(self.entity_data.entity.supported_features.value)
```

Comment:
`.value` shouldn't be needed here.

Reply:
Mirrored what was done in esphome and addressed in all platforms that were doing this in ZHA